TYPO3.Flow/Classes/TYPO3/Flow/Package/Package.php

```diff
@@ -425,97 +417,159 @@ public function getDocumentationPath()
-     * @param string $subDirectory Used internally
-     * @param integer $recursionLevel Used internally
-     * @return array
-     * @throws \TYPO3\Flow\Package\Exception if recursion into directories was too deep or another error occurred
+     * @return string
+     * @api
+     * TODO: Should be added to the interface in the next major Flow version (4.0)
      */
-    protected function buildArrayOfClassFiles($classesPath, $extraNamespaceSegment = '', $subDirectory = '', $recursionLevel = 0)
+    public function getInstalledVersion()
     {
-        $classFiles = array();
-        $currentPath = $classesPath . $subDirectory;
-        $currentRelativePath = substr($currentPath, strlen($this->packagePath));
-        $namespacePrefix = '';
+        return PackageManager::getPackageVersion($this->composerName);
+    }
 
-        if ($this->getAutoloadType() === self::AUTOLOADER_TYPE_PSR4) {
-            $namespacePrefix = $this->getNamespace();
+    /**
+     * @param string $autoloadType
+     * @param string $autoloadNamespace
+     * @param string $autoloadPath
+     * @return string
+     */
+    protected function normalizeAutoloadPath($autoloadType, $autoloadNamespace, $autoloadPath)
+    {
+        $normalizedAutoloadPath = $autoloadPath;
+        if ($autoloadType === ClassLoader::MAPPING_TYPE_PSR0) {
+            $normalizedAutoloadPath = Files::concatenatePaths([
+                    $autoloadPath,
+                    str_replace('\\', '/', $autoloadNamespace)
+                ]) . '/';
         }
 
-        if (!is_dir($currentPath)) {
-            return array();
-        }
-        if ($recursionLevel > 100) {
-            throw new Exception('Recursion too deep while collecting class files.', 1166635495);
+        return $normalizedAutoloadPath;
+    }
+
+    /**
+     * @param string $baseAutoloadPath
+     * @param string $autoloadNamespace
+     * @return \Generator
+     */
+    protected function getClassesInNormalizedAutoloadPath($baseAutoloadPath, $autoloadNamespace)
+    {
+        $autoloadNamespace = trim($autoloadNamespace, '\\') . '\\';
+        $directories = [''];
+        while ($directories !== []) {
+            $currentRelativeDirectory = array_pop($directories);
+            $currentAbsoluteDirectory = $baseAutoloadPath . $currentRelativeDirectory;
+            if ($handle = opendir($currentAbsoluteDirectory)) {
+                while (false !== ($filename = readdir($handle))) {
+                    if ($filename[0] === '.') {
+                        continue;
+                    }
+                    $pathAndFilename = $currentAbsoluteDirectory . $filename;
+                    if (is_dir($pathAndFilename)) {
+                        $directories[] = $currentRelativeDirectory . $filename . '/';
+                        continue;
+                    }
+                    if (strpos(strrev($filename), 'php.') === 0) {
```

Comment:
Isn't that slower apart from being harder to read than "substr($filename, -4[, 4]) === '.php'"?